File: cpp/generate_parenthesis.cpp
// Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

 

// Example 1:

// Input: n = 3
// Output: ["((()))","(()())","(())()","()(())","()()()"]
// Example 2:

// Input: n = 1
// Output: ["()"]

class Solution {
public:
    vector<string> generator (int i, int j, int n, vector<string> & ans, string &res){
        if (i ==j == n) {
            return ans.push_back(res);
        }
        if (i < n){
            res.push_back('(');
            generator(i+1, j, n, ans, res);
        }
        if (j<i){
            res.push_back(')');
            generator(i, j+1, n, ans, res);
        } 
    }
    vector<string> generateParenthesis(int n) {
        int i, j;
        vector<string> ans, res;
        string s;

        res = generator(i, j, n, ans, s);
        return res;
    }
};

File: python/generate_parenthesis.py
# Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

 

# Example 1:

# Input: n = 3
# Output: ["((()))","(()())","(())()","()(())","()()()"]
# Example 2:

# Input: n = 1
# Output: ["()"]

class Solution:
    def generator(self,i,j, n, res, ans):
        if i == j == n :
            ans.append(res)
            return
        if i < n:
            self.generator(i+1, j, n, res + "(", ans)
        if j < i:
            self.generator(i, j+1, n ,res+ ")", ans)
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        self.generator(0,0,n, "", ans)
        return ans